### PR TITLE
plpgsql/parser: fully parse simple PL/pgSQL statements

### PR DESCRIPTION
**Pull request by DrewKimball**
#### plpgsql/parser: fix number parsing for PL/pgSQL

This commit fixes a bug in the PL/pgSQL parser where numbers would be
scanned using the SQL token IDs instead of the PL/pgSQL token IDs,
which could prevent expressions with numbers from being parsed correctly.

#### plpgsql/parser: add parser support for simple PL/pgSQL statements

This patch adds support in the PL/pgSQL parser for parsing `RETURN`, `IF`,
simple `LOOP`, assignment, and variable declaration statements. This includes
parsing SQL expressions into `PLpgSQLExpr` form. A future PR will add
optbuilder support for these statements.

Release note: None

Epic: [CRDB-799](https://cockroachlabs.atlassian.net/browse/CRDB-799)

### CODE REVIEW

File: pkg/sql/plpgsql/parser/plpgsql.y [hunk 1]
```diff
@@ -335,60 +351,67 @@ opt_semi:
...
   {
-  // TODO(chengxiong): log declare section ($1) here.
-  header := $1.plpgsqlDeclareheader()
-  stmtBlock := &plpgsqltree.PLpgSQLStmtBlock{
-    Label: header.label,
-    Body: $3.plpgsqlStatements(),
-  }
-  if header.initVars != nil {
-    stmtBlock.InitVars = header.initVars
+    // TODO(chengxiong): log declare section ($1) here.
```
Comment (chengxiong-ruan) [inline pkg/sql/plpgsql/parser/plpgsql.y:356]: I forgot why this todo was added, I think we can remove it

File: pkg/sql/plpgsql/parser/utils.go [hunk 2]
```diff
@@ -9,10 +9,3 @@
 // licenses/APL.txt.
 
 package parser
-
```
Comment (chengxiong-ruan) [inline pkg/sql/plpgsql/parser/utils.go:12]: this file can be removed for now.

File: pkg/sql/sem/plpgsqltree/statements.go [hunk 3]
```diff
@@ -57,23 +57,26 @@ func (s *PLpgSQLStatementImpl) plpgsqlStmt() {}
 type PLpgSQLStmtBlock struct {
 	PLpgSQLStatementImpl
 	Label      string
+	Decs       []PLpgSQLDecl
```
Comment (chengxiong-ruan) [inline pkg/sql/sem/plpgsqltree/statements.go:60]: `Decs` -> `Decls`?

**Review by chengxiong-ruan — APPROVED** [review]
just a few nits. Thanks for working on this!
also there were a few more statement parsing (including `stmt_for`) implemented in [this PR](https://github.com/cockroachdb/cockroach/pull/101246) hope can help as well.

**Review by mgartner — APPROVED** [review]
:lgtm: modulo @chengxiong-ruan's comments

> Reviewed 1 of 1 files at r1, 16 of 16 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/104117#-:-NWnX2eZ6DocQYMlTZvM:b-dykk36)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @DrewKimball, @rharding6373, and @ZhouXing19)

**Review by rharding6373 — APPROVED** [review]
:lgtm: Nice work!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/104117#-:-NWrrQK8BCNHvLd3vHyW:bcmw7ss)* status: :shipit: complete! 2 of 0 LGTMs obtained (waiting on @DrewKimball and @ZhouXing19)

___
*[`pkg/sql/sem/plpgsqltree/statements.go` line 179 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/104117#-NWrqBvX1oWpPNcb6znU:-NWrqBvX1oWpPNcb6znV:b-35df6u) ([raw file](https://github.com/cockroachdb/cockroach/blob/b337e3d086b715f7c9d3401e0b07ca7edc59a9b6/pkg/sql/sem/plpgsqltree/statements.go#L179)):*
> ```Go
> 	}
> 	ctx.WriteString("END IF\n")
> 	ctx.WriteString("<NOT DONE YET>\n")
> ```

Do you know why we print "<NOT DONE YET>" here? Seems to be added to Format() inconsistently for the different statement types.

**Review by DrewKimball — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/104117)* status: :shipit: complete! 2 of 0 LGTMs obtained (waiting on @rharding6373 and @ZhouXing19)

___
*[`pkg/sql/sem/plpgsqltree/statements.go` line 179 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/104117#-NWrqBvX1oWpPNcb6znU:-NWry3N_9hDvzKlYn96c:bk4lqyv) ([raw file](https://github.com/cockroachdb/cockroach/blob/b337e3d086b715f7c9d3401e0b07ca7edc59a9b6/pkg/sql/sem/plpgsqltree/statements.go#L179)):*
<details><summary><i>Previously, rharding6373 (Rachael Harding) wrote…</i></summary><blockquote>

Do you know why we print "<NOT DONE YET>" here? Seems to be added to Format() inconsistently for the different statement types.
</blockquote></details>

Do you mean why we print `<NOT DONE YET>`?

File: pkg/sql/sem/plpgsqltree/statements.go (hunk 3 above)
Comment (DrewKimball) [reply pkg/sql/sem/plpgsqltree/statements.go]: Done.

File: pkg/sql/plpgsql/parser/utils.go (hunk 2 above)
Comment (DrewKimball) [reply pkg/sql/plpgsql/parser/utils.go:12]: Done.

File: pkg/sql/plpgsql/parser/plpgsql.y (hunk 1 above)
Comment (DrewKimball) [reply pkg/sql/plpgsql/parser/plpgsql.y]: Done.

**Review by rharding6373 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/104117)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @chengxiong-ruan, @DrewKimball, @mgartner, and @ZhouXing19)

___
*[`pkg/sql/sem/plpgsqltree/statements.go` line 179 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/104117#-NWrqBvX1oWpPNcb6znU:-NWs13pECUtWNOcDGofz:brsix5p) ([raw file](https://github.com/cockroachdb/cockroach/blob/b337e3d086b715f7c9d3401e0b07ca7edc59a9b6/pkg/sql/sem/plpgsqltree/statements.go#L179)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

Do you mean why we print `<NOT DONE YET>`?
</blockquote></details>

Yes, sorry. And as a corollary, why do we no longer print it for EXIT?

**Review by DrewKimball — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/104117)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @chengxiong-ruan, @mgartner, @rharding6373, and @ZhouXing19)

___
*[`pkg/sql/sem/plpgsqltree/statements.go` line 179 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/104117#-NWrqBvX1oWpPNcb6znU:-NWs2vD--LXMGuUL7eGy:bvv822g) ([raw file](https://github.com/cockroachdb/cockroach/blob/b337e3d086b715f7c9d3401e0b07ca7edc59a9b6/pkg/sql/sem/plpgsqltree/statements.go#L179)):*
<details><summary><i>Previously, rharding6373 (Rachael Harding) wrote…</i></summary><blockquote>

Yes, sorry. And as a corollary, why do we no longer print it for EXIT?
</blockquote></details>

It's because those format methods aren't finished, and I removed for EXIT because I finished it.

**Comment by DrewKimball** [comment]
Just fixed a bug where `CONTINUE` was being parsed into `PLpgSQLStmtExit` instead of `PLpgSQLStmtContinue`.

**Comment by DrewKimball** [comment]
TFYRs!

bors r+

**Comment by DrewKimball** [comment]
Bors failure looks like:
```
Updating sources: agent side checkout
  Full checkout enforced. Reason: ["Delete all files before the build" turned on]
  VCS Root: Cockroach (All Pull Requests)
    revision: 95362b159e0fdfad9eda049bbde130e555a83f87
    Git version: 2.39.0.0
    Update git mirror (/home/agent/system/git/cockroach.git)
      /usr/bin/git config http.sslCAInfo
      /usr/bin/git show-ref
      /usr/bin/git -c credential.helper= ls-remote origin
      Failed to resolve artifact dependency <Public/External / Cockroach / CI / Builds / Build Docker Image (Linux x86_64), build #67538 [id 10408882]>: Failed to download file 'cockroach-docker-image-nonroot.tar.gz': Failed to download [https://teamcity.cockroachdb.com/httpAuth/repository/download/Cockroach_ScratchProjectPutTcExperimentsInHere_JaneDockerImage/10408882.tcbuildid/cockroach-docker-image-nonroot.tar.gz]: java.io.IOException: Illegal status [500] while downloading https://teamcity.cockroachdb.com/httpAuth/repository/download/Cockroach_ScratchProjectPutTcExperimentsInHere_JaneDockerImage/10408882.tcbuildid/cockroach-docker-image-nonroot.tar.gz: Internal Server Error (jetbrains.buildServer.artifacts.impl.SourcePathAwareResolvingFailedException)
      Failed to resolve artifacts from <Public/External / Cockroach / CI / Builds / Build Docker Image (Linux x86_64), build #67538 [id 10408882]>
      Completed the Golang tests processing.
      Golang tests not found.
      Build failed to start. Artifacts will not be published for this build
```
doesn't look related.

bors retry

**Comment by DrewKimball** [comment]
bors retry

**Pull request merged**